core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+
+  /**
+   * The size of the block (in bytes)
+   */
+  var size: Long = 0
+
+  /**
+   * The number of times that this block has been locked for reading.
+   */
+  var readerCount: Int = 0
+
+  /**
+   * The task attempt id of the task which currently holds the write lock for this block, or -1
+   * if this block is not locked for writing.
+   */
+  var writerTask: Long = -1
+
+  // Invariants:
+  //     (writerTask != -1) implies (readerCount == 0)
+  //     (readerCount != 0) implies (writerTask == -1)
+  // TODO: add assertions around every method
+
+  /**
+   * True if this block has been removed from the BlockManager and false otherwise.
+   * This field is used to communicate block deletion to blocked readers / writers (see its usage
+   * in [[BlockInfoManager]]).
+   */
+  var removed: Boolean = false
+
+  // TODO: Add timestamps on lock acquisitions
+}
+// In debugging mode, check that locks haven't been held for too long.
+// Every few minutes, dump debug info.
+
+/**
+ * Component of the [[BlockManager]] which tracks metadata for blocks and manages block locking.
+ *
+ * The locking interface exposed by this class is readers-writers lock. Every lock acquisition is
+ * automatically associated with a running task and locks are automatically released upon task
+ * completion or failure.
+ *
+ * This class is thread-safe.
+ */
+private[storage] class BlockInfoManager extends Logging {
+
+  private type TaskAttemptId = Long
+
+  /**
+   * Used to look up metadata for individual blocks. Entries are added to this map via an atomic
+   * set-if-not-exists operation ([[lockNewBlockForWriting()]]) and are removed
+   * by [[removeBlock()]].
+   */
+  @GuardedBy("this")
+  private[this] val infos = new mutable.HashMap[BlockId, BlockInfo]
+
+  /**
+   * Tracks the set of blocks that each task has locked for writing.
+   */
+  @GuardedBy("this")
+  private[this] val writeLocksByTask =
+    new mutable.HashMap[TaskAttemptId, mutable.Set[BlockId]]
+      with mutable.MultiMap[TaskAttemptId, BlockId]
+
+  /**
+   * Tracks the set of blocks that each task has locked for reading, along with the number of times
+   * that a block has been locked (since our read locks are re-entrant). This is thread-safe.
+   */
+  private[this] val readLocksByTask: LoadingCache[lang.Long, ConcurrentHashMultiset[BlockId]] = {
+    // We need to explicitly box as java.lang.Long to avoid a type mismatch error:
+    val loader = new CacheLoader[java.lang.Long, ConcurrentHashMultiset[BlockId]] {
+      override def load(t: java.lang.Long) = ConcurrentHashMultiset.create[BlockId]()
+    }
+    CacheBuilder.newBuilder().build(loader)
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  /**
+   * Returns the current tasks's task attempt id (which uniquely identifies the task), or -1024
+   * if called outside of a task (-1024 was chosen because it's different than the -1 which is used
```

Comment:
can you make these `-1` and `-1024` constants somewhere, like `val NO_WRITER = -1` or something